frontend/packages/citizen-frontend/src/applications/editor/FileUpload.tsx

```diff
@@ -0,0 +1,342 @@
+      return faFileWord
+    default:
+      return faFile
+  }
+}
+
+const inProgress = (file: FileObject): boolean =>
+  !file.error && file.progress !== 100
+
+export default React.memo(function FileUpload({
+  files,
+  onUpload,
+  onDelete
+}: FileUploadProps) {
+  const t = useTranslation()
+
+  const [uploadedFiles, setUploadedFiles] = useState<FileObject[]>(
+    files.map(attachmentToFile)
+  )
+
+  const onChange = (event: React.ChangeEvent<HTMLInputElement>) => {
+    // FileList is empty when one file is uploaded and then selecting another one is cancelled,
+    // clearing the input's value
+    if (event.target.files && event.target.files.length > 0) {
+      void addAttachment(event.target.files[0], onUpload)
+    }
+  }
+
+  const onDrop = (event: React.DragEvent<HTMLLabelElement>) => {
+    if (event.dataTransfer && event.dataTransfer.files[0]) {
+      void addAttachment(event.dataTransfer.files[0], onUpload)
+    }
+  }
+
+  const deleteFile = async (id: UUID) => {
+    try {
+      const result = await onDelete(id)
+      result.isSuccess &&
+        setUploadedFiles((old: FileObject[]) => {
+          return old.filter((item) => item.id !== id)
+        })
+    } catch (e) {
+      console.error(e)
+    }
+  }
+
+  const addAttachment = async (
+    file: File,
+    onUpload: (
+      file: File,
+      onUploadProgress: (progressEvent: ProgressEvent) => void
+    ) => Promise<Result<UUID>>
+  ) => {
+    const error = file.size > 10000000 ? 'file-too-big' : undefined
+    const fileObject: FileObject = {
+      id: Math.random().toString(),
+      key: 1,
+      file,
+      name: file.name,
+      contentType: file.type,
+      progress: 0,
+      error
+    }
+
+    if (error) {
+      return
+    }
+
+    const updateProgress = ({ loaded, total }: ProgressEvent) => {
+      const progress = Math.round((loaded / total) * 100)
+      setUploadedFiles((old: FileObject[]) => {
+        const others = old.filter((item) => item.id !== fileObject.id)
+        const progressedFile = { ...fileObject, progress }
+        return [...others, progressedFile]
+      })
+    }
+
+    try {
+      const result = await onUpload(file, updateProgress)
+      result.isSuccess &&
+        setUploadedFiles((old: FileObject[]) => {
+          const others = old.filter((item) => item.id !== fileObject.id)
+          const progressedFile = {
+            ...fileObject,
+            progress: 100,
+            id: result.value
+          }
+          return [...others, progressedFile]
+        })
+    } catch (e) {
+      console.error(e)
+      setUploadedFiles((old: FileObject[]) => {
+        const others = old.filter((item) => item.id !== fileObject.id)
+        const progressedFile = { ...fileObject, error: 'server-error' }
+        return [...others, progressedFile]
+      })
+    }
+  }
+
+  return (
+    <FileUploadContainer>
+      <FileInputLabel className="file-input-label" onDrop={onDrop}>
```

Comment:
`input` and `h4` inside `label` sounds kinda exotic. Not sure if that may cause accessibility issues or such.

Reply:
I think `input` is quite often seen inside `label` but this `h4` is truly a funky one.